hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/AbstractHoodieTableFileIndex.scala

```diff
@@ -0,0 +1,309 @@
+import scala.collection.mutable
+
+/**
+ * A file index which support partition prune for hoodie snapshot and read-optimized query.
+ *
+ * Main steps to get the file list for query:
+ * 1、Load all files and partition values from the table path.
+ * 2、Do the partition prune by the partition filter condition.
+ *
+ * There are 3 cases for this:
+ * 1、If the partition columns size is equal to the actually partition path level, we
+ * read it as partitioned table.(e.g partition column is "dt", the partition path is "2021-03-10")
+ *
+ * 2、If the partition columns size is not equal to the partition path level, but the partition
+ * column size is "1" (e.g. partition column is "dt", but the partition path is "2021/03/10"
+ * who's directory level is 3).We can still read it as a partitioned table. We will mapping the
+ * partition path (e.g. 2021/03/10) to the only partition column (e.g. "dt").
+ *
+ * 3、Else the the partition columns size is not equal to the partition directory level and the
+ * size is great than "1" (e.g. partition column is "dt,hh", the partition path is "2021/03/10/12"),
+ * we read it as a Non-Partitioned table because we cannot know how to mapping the partition
+ * path with the partition columns in this case.
+ *
+ */
+abstract class AbstractHoodieTableFileIndex(engineContext: HoodieEngineContext,
+                                            metaClient: HoodieTableMetaClient,
+                                            configProperties: TypedProperties,
+                                            specifiedQueryInstant: Option[String] = None,
+                                            @transient fileStatusCache: FileStatusCacheTrait) {
+  /**
+   * Get all completeCommits.
+   */
+  lazy val completedCommits = metaClient.getCommitsTimeline
+    .filterCompletedInstants().getInstants.iterator().toList.map(_.getTimestamp)
+  /**
+   * Get the partition schema from the hoodie.properties.
+   */
+  private lazy val _partitionColumns: Array[String] =
+    metaClient.getTableConfig.getPartitionFields.orElse(Array[String]())
+
+  private lazy val fileSystemStorageConfig = FileSystemViewStorageConfig.newBuilder()
+    .fromProperties(configProperties)
+    .build()
+  private lazy val metadataConfig = HoodieMetadataConfig.newBuilder
+    .fromProperties(configProperties)
+    .build()
+  protected val basePath: String = metaClient.getBasePath
+
+  private val queryType = configProperties(QUERY_TYPE.key())
+  private val tableType = metaClient.getTableType
+
+  @transient private val queryPath = new Path(configProperties.getOrElse("path", "'path' option required"))
+  @transient
+  @volatile protected var cachedFileSize: Long = 0L
+  @transient
+  @volatile protected var cachedAllInputFileSlices: Map[PartitionPath, Seq[FileSlice]] = _
+  @volatile protected var queryAsNonePartitionedTable: Boolean = _
+  @transient
+  @volatile private var fileSystemView: HoodieTableFileSystemView = _
+
+  refresh0()
```

Comment:
I assume most of these logic are copied over from `HoodieFileIndex`.

Reply:
Correct, it's cherry-picked from `HoodieFileIndex` and neutralized to be engine-agnostic